### ISSUE TITLE
a swapchains imageExtent must match the windows size

### ISSUE THREAD

**Issue by nco2k** [KhronosGroup/Vulkan-Docs#484]
i hope im writing to the right repo.

the vulkan spec says in section 30.2.4. 'Win32 Platform': "With Win32, minImageExtent, maxImageExtent, and currentExtent are the window size. Therefore, a swapchain’s imageExtent must match the window’s size."

i honestly dont see a reason for this artificial limitation. why does the 'imageExtent' have to match the 'currentExtent'?

i am letting the user select two different resolutions. one for the windowed mode and one for the (borderless window) fullscreen mode. if the windowed mode is for example 1280x720 and the fullscreen mode is for example 1024x768, then i use the bigger values and create a surface of 1280x768. the reason i do this is to quickly being able to switch between the windowed and (borderless window) fullscreen mode, without the need to unnecessarily recreate the whole swapchain etc.

with my fictional example from above, the data would look as following on my gtx 1060:

VkSurfaceCapabilitiesKHR::currentExtent.width = 1280
VkSurfaceCapabilitiesKHR::currentExtent.height = 720

VkSurfaceCapabilitiesKHR::minImageExtent.width = 1
VkSurfaceCapabilitiesKHR::minImageExtent.height = 1

VkSurfaceCapabilitiesKHR::maxImageExtent.width = 32768
VkSurfaceCapabilitiesKHR::maxImageExtent.height = 32768

VkSwapchainCreateInfoKHR::imageExtent.width = 1280
VkSwapchainCreateInfoKHR::imageExtent.height = 768

my selected 'imageExtent' is larger than 'currentExtent', but in the range between 'minImageExtent' and 'maxImageExtent'. i dont plan on drawing past the 'currentExtent' anyway. i just want to avoid unnecessarily recreating the whole swapchain.

we have viewports, we have scissors, so i dont know why we cant create a swapchain image larger than 'currentExtent'. despite getting the "vkCreateSwapChainKHR() called with imageExtent which is not equal to the currentExtent." validation error message, creating a larger swapchain image actually works flawless on my win7 system.

i honestly think that this should be allowed. constantly recreating the whole swapchain while/after resizing a window is very inefficient. if we could officially use a larger 'ImageExtent', then we would have to recreate the swapchain only when the 'currentExtent' is larger than the 'ImageExtent'. when its smaller or equal, we dont.

not to mention that we could also process the WM_GETMINMAXINFO message on windows and set a maximum allowed size for our window. then we could create a swapchain image right away, that matches those limitations and we could resize the window as often as we like, without ever having to recreate the swapchain. unless we get an error like 'VK_ERROR_OUT_OF_DATE_KHR' of course.

either way, i hope you consider this to be changed and officially supported. thank you for your time.

**Comment by krOoze** [comment]
> i honestly dont see a reason for this artificial limitation. why does the 'imageExtent' have to match the 'currentExtent'?

Because possibly on Windows it is not obvious what should happen when you swap an image of different size than the window &mdash; currently it seems to create black borders or snip the image to fit, which is quite likely not what the user wants to see.

> i am letting the user select two different resolutions. one for the windowed mode and one for the (borderless window) fullscreen mode. if the windowed mode is for example 1280x720 and the fullscreen mode is for example 1024x768, then i use the bigger values and create a surface of 1280x768. the reason i do this is to quickly being able to switch between the windowed and (borderless window) fullscreen mode, without the need to unnecessarily recreate the whole swapchain etc.

It is not "borderless fullscreen" if the fullscreen resolution differs from the current desktop resolution.

Recreating Vulkan swapchain is quick enough in my experience. It is a glorified realloc really.

> i honestly think that this should be allowed. constantly recreating the whole swapchain while/after resizing a window is very inefficient.

BTW it actually work the way you say. Unless you get 'VK_ERROR_OUT_OF_DATE_KHR'.

It's not like some OS provided scaling is necessarily an efficient thing to do. It probably needs to allocate another buffer for the scaled image and whatnot. The system possibly needs to create another swapchain behind your back anyway in that case.

**Comment by nco2k** [comment]
> Because possibly on Windows it is not obvious what should happen when you swap an image of different size than the window.

thats why im asking for a proper implementation of this. that way you will know what should and will happen if you decide to use it like that. but currently you get validation errors if you decide to use it like that, because its not officially supported.

if you have a window of 500x500 and a swapchain image of 600x600, you set the scissors and viewports to match your window and when you swap the image, everything larger than 500x500 should be discarded. the swapchain shouldnt care wether you have a 500x500 image or a 9000x9000 image. all it should care about is the part you actually use.

> It is not "borderless fullscreen" if the fullscreen resolution differs from the current desktop resolution.

thats why im using ChangeDisplaySettingsEx() to temporary change the desktop resolution and match the users selected resolution. many companies and emulators had been using this method for more than 15 years now, in order to avoid having to use an exclusive fullscreen mode. using a borderless fullscreen window instead of an exclusive fullscreen mode, does have some advantages after all.

but there are some companies like a big company that starts with the letter B, that offers a borderless fullscreen window, but doesnt let you change the resolution or even allow you to do a proper taskswitch when running it. which is a perfect example on how not to use a borderless fullscreen window. but this has nothing to do with the actual topic.

> Recreating Vulkan swapchain is quick enough in my experience.

its not about speed, its about the why. there is no technical reason for this limitation.

> You need to recreate your cmdbuffers.

if larger swapchains were officially supported, then you wouldnt have to re-create the cmdbuffers etc., because the swapchain never changed in the first place.

> BTW it actually work the way you say.

well no, when the window size changes, you have to make sure that the swapchain size matches the window size. if you dont, you will get the usual "a swapchains imageExtent must match the windows size" validation error message.

and again, this is not about speed or whatever. it just feels like an artficial limitation that shouldnt be here in the first place.

**Comment by ianelliottus** [comment]
> > Because possibly on Windows it is not obvious what should happen when you swap an image of different size than the window.

> im asking for a proper implementation of this. that way you will know what should and will happen if you decide to use it like that. but currently you get validation errors if you decide to use it like that, because its not officially supported.

As I explained in a different forum, you are benefiting from a driver bug, that they plan to fix in a future release.  I believe that @krOoze is correct.  Note that the same valid usage rules exist for both Windows and Linux.

That said, I've asked for a discussion of your request to see if Khronos wants an explicit change to the spec.

**Comment by nco2k** [comment]
> That said, I've asked for a discussion of your request to see if Khronos wants an explicit change to the spec.

thanks a lot! much appreciated.

**Comment by krOoze** [comment]
> Note that the same valid usage rules exist for both Windows and Linux.

Well, for X. Wayland and Mir are different. And Android is technically Linux too 😛 .

> That said, I've asked for a discussion of your request to see if Khronos wants an explicit change to the spec.

Can't hurt to ask. I would not hold much hope though.
My bet is they say it is against Vulkan philosophy, because the Implementation would have to code it, if OS does not support it directly and the Application can implement scaling itself if it wishes so (using `vkCmdBlit`).

**Comment by ianelliottus** [comment]
I apologize @nco2k and @krOoze that the person who was going to report back has been busy and obviously hasn't reported back yet.

We did discuss this within Khronos, right after I said I'd bring it up.  They decided that they did not want to change the spec per your request.  Various reasons, which I don't remember off the top of my head (sorry, but in looking through my browser tabs, I saw that nobody responded, and I at least wanted to get back to you with the bottom line).  Please re-open and respond if you want me to ping the person who agreed to get back with you.

**Comment by michaeljclark** [comment]
> we have viewports, we have scissors, so i dont know why we cant create a swapchain image larger than 'currentExtent'. despite getting the "vkCreateSwapChainKHR() called with imageExtent which is not equal to the currentExtent." validation error message, creating a larger swapchain image actually works flawless on my win7 system.

I agree. Mandating that the swapchain image extent must match the surface image extent goes against prior practice with OpenGL, at least on Linux, and introduces (or simplifies?) synchronization issues during window resizing.

### Disadvantages of allowing `(imageExtent != currentExtent)`
understand why one would want to resize buffers synchronously within the resize event, i.e. xlib.ResizeWindow with Vulkan on X11. If the buffer resizes are done synchronously inside of the resize event, then revealing an undrawn extent of the window background color can be avoided. The OS compositor can make sure the application has rendered a frame with the new size before compositing new frames during resize.

### Advantages of allowing `(imageExtent != currentExtent)`
understand why one may wish to do this asynchronously. OpenGL on X11 allows both behaviors. Some apps resize synchronously but they also tend to have a more sluggish response to resize the window frame than ones that do so asynchronously. apps that resize asynchronously can reveal an undrawn extent of the window background color.

### Asynchronous resizes

I am porting a GLFW app from OpenGL to Vulkan and I am experiencing some race conditions during window resize with GLFW which is why I am here. It might be a GLFW bug because the GLFW vulkan test segfaults when I try to resize a window.  It is quite a heavy operation to perform synchronously during the resize event, as there may be multiple frames before the new size makes it to the front buffer. I notice that [island](https://github.com/tgfrerer/island.git) resizes its windows synchronously, but again, it is quite sluggish to resize the window on Linux. I also think there are some relatively complex synchronization issues when there are many buffers that need to be recreated, such as when multisampling is enabled. _i.e._

- `num_images = swapchain images × samples + depth buffer + resolve buffer`

The goal would be to allow asynchronous rendering and resizing in parallel with multiple swapchain images and sample buffers. Not a simple task. The resulting spec decision possibly also needs to be reconciled against these extensions:
- [EGL_KHR_partial_update](https://www.khronos.org/registry/EGL/extensions/KHR/EGL_KHR_partial_update.txt)
- [EGL_KHR_swap_buffers_with_damage](https://www.khronos.org/registry/EGL/extensions/KHR/EGL_KHR_swap_buffers_with_damage.txt)

One possibility would be to define a new extension in the same vein as [VK_KHR_swapchain_mutable_format](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_KHR_swapchain_mutable_format.html) perhaps _VK_KHR_swapchain_mutable_extent_ with the text _allows creation of a swapchain with an imageExtent that differs from the surface currentExtent…_ and adds the following info bits for `vkCreateSwapchainKHR` and `vkCreateImage` respectively:
- _VkSwapchainCreateFlagBitsKHR_ `VK_IMAGE_CREATE_MUTABLE_EXTENT_BIT_KHR`
- _VkImageCreateFlagBits_ `VK_SWAPCHAIN_CREATE_MUTABLE_EXTENT_BIT_KHR`

### Tiled resizes

There is also the issue of tiling with [VkImageTiling](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkImageTiling.html) and [VkImageDrmFormatModifierListCreateInfoEXT](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkImageDrmFormatModifierListCreateInfoEXT.html) where an implementation scatter-gathers an image from a list of smaller images of some implementation-defined size quanta.

I have been thinking about an optimal implementation for butter-smooth window resize and found a reference to tiled images in the open-source intel drm driver. The idea is that one has tiles pre-allocated so that an implementation can write to them immediately without any new allocation, keeping margin so that as new areas are revealed during resizing there is already memory allocated that can be drawn into practically immediately. This is somewhat of a deep implementation detail, but it would also depend on the _swapchain.imageExtent_ differing from _surface.currentExtent_.

With multiple swap chain images, the size change essentially rolls across multiple frames with several queue submits and present calls. Doing that without frame drops while say, also composting video, sounds quite challenging. That is my hope.

On Linux of course…

**Comment by ShabbyX** [comment]
FYI, the new [VK_EXT_surface_maintenance1][surface] and [VK_EXT_swapchain_maintenance1][swapchain] extensions address this issue. Please see this [blog post][blog] for a summary.

[surface]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_surface_maintenance1.html
[swapchain]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_swapchain_maintenance1.html
[blog]: https://www.khronos.org/blog/resolving-longstanding-issues-with-wsi

TL;DR: it is now possible to create a swapchain that doesn't match the surface extents (if supported)
